feeds/keystore-generator.sh

```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+
+# Tool to generate new encrypted keystore with Ethereum address matching
+# a specific first byte identifier.
+#
+# Dependencies:
+# - cast, see https://getfoundry.sh
+# - unix utilities
+#
+# Usage:
+#
+#  $ ./key-generator.sh <0x prefixed byte> <keystore path> <keystore password>
+#
+# Example:
+#
+#  $ ./key-generator.sh 0xff ./keystores test
+set -euo pipefail # Enable strict mode for bash
+
+# Fail if foundry toolchain's cast not installed.
+if ! command -v cast &> /dev/null
+then
+  echo "Error: Please install the foundry toolchain's cast tool, see https://getfoundry.sh"
+fi
+
+# Fail if invalid number of arguments provided.
+if [ "$#" -ne 3 ]; then
+  echo "Usage: $0 <0x prefixed byte> <keystore path> <keystore password>"
+  exit 1
+fi
+
+# Store arguments into variables.
+assigned_id="$1"
```

Comment:
do we want an id check to ensure the user doesn't grab an id that is already taken? I added it here (per the convo on the notion doc):

https://gist.github.com/jar-o/e062c0a9cc8697866973f6d8b35cab36

note that it does query the greetings endpoint, not sure if that matters or not

Reply:
Ahh damn! I just copied the script - forgot about your update, sorry.

On the other hand I think this is worth a broader discussion: Do we want to assign ids to validator or do we wanna keep them random (but unique/valid)?

Some stuff I can think of:

Pro _assign ids_:
- In case of a private key leak, eg due to a validator's host hacked, cryptographic issues, ..., we may need to have (multiple) validators create new keys at the same time. In this case the random method is not always correct, ie two validators could end up with new addresses that have same id
- We could encode extra data into the ids (eg only `0xF_` validators provide proprietary data X) and/or order them (eg start at 0x00 and just count upwards). Think for example how opcodes are categorised into ranges, see [evm opcodes](https://github.com/ethereum/go-ethereum/blob/master/core/vm/opcodes.go#L31). Note that address encoding is kinda common onchain, eg Uniswap v4 will invoke different hooks depending on bit masks encoded into the address, see [here](https://github.com/Uniswap/v4-core/blob/main/src/libraries/Hooks.sol#L14-L17).

Pro _random ids_:
- ???

Note that the `greetings-api` cannot be the source of truth neither. If we onboard more than one validator at a time, their ids may collide again.

Reply:
(_just to be clear, my variation makes no attempt to assign ids, it just checks to make sure what the validator picked is not already taken_)

I think letting validators pick means a higher than average chance of getting two to pick the same id since humans suck at entropy (`42`, `69`, `aa`...). random ids with a check against greetings api to discard one that's taken is probably the best probability we'll see.

but assigning ids is the only way to make sure there are no collisions, which imo boils down to the appetite of the validators. I mean we could say "here's a script to generate your key", use this `id`". as long as they understood we have no way of seeing their private key, maybe they don't care  _shrug_.

Reply:
> a check against greetings api

This check is not sufficient. We plan to onboard validators in batches but the greetings api is only updated once we start the onboarding, ie after the key generation.

> letting validators pick means a higher than average chance of getting two to pick the same id

Agree, lets just decide internally on an assignment scheme and roll with it. Probably best is just to start at `0x00` and increment(?)

> as long as they understand we have no way of knowing their private key

This is the reason for the _simple_ and easy to audit shell script. The actual key generation is happening through `cast`, a tool we don't control, and the shell script is "as easy" as possible.

Obv for someone without any programming skills its still a blackbox though. I would propose to make the script public and ask for a "review" from the broader community to establish trust for the validators. However, lets see whether thats necessary first, ie whether validators mention trust issues.

Reply:
> Agree, lets just decide internally on an assignment scheme and roll with it. Probably best is just to start at 0x00 and increment(?)

🤝